apps/prairielearn/src/batched-migrations/20250318174636_course_instance_usages__submissions__backfill_v3.sql

```diff
@@ -0,0 +1,60 @@
+-- BLOCK delete_old_usages
+DELETE FROM course_instance_usages
+WHERE
+  type = 'Submission'
+  AND date >= $START_DATE
+  AND date < $END_DATE;
+
+-- BLOCK select_bounds
+SELECT
+  min(id),
+  max(id)
+FROM
+  submissions
+WHERE
+  date >= $START_DATE
+  AND date < $END_DATE;
```

Comment:
This query takes about 25 minutes in prod. The problem is that the plan uses a `Index Scan using submissions_pkey on submissions`. In contrast, doing a `count(*)` with the same WHERE conditions results in a plan that uses `Parallel Index Only Scan using submissions_date_idx on submissions`. I can't figure out a way to get the `select_bounds` query to use the `submissions_date_idx` index.

An alternative would be to use `min: 1` and get the `max(id)` from the whole table (that is fast, because it uses an index only scan). Then we could shift the `date` condition into the backfill query itself. This would mean that we iterative over all submission rows and just fail to do anything with most of them. This would be slow but safe because it's running in batches.

Reply:
If you split this into two queries, you can write them such that Postgres sees it can use the index. For instance, for the lower bound:

```sql
SELECT id FROM submissions WHERE date >= '2025-02-15T00:00:00Z' ORDER BY date ASC, id ASC LIMIT 1;
```

This runs in ~0.2ms. You can do something similar for the upper bound.

Reply:
Nice! Changed in https://github.com/PrairieLearn/PrairieLearn/pull/11598/commits/8640721a61cd52ed8f1cc37fd95fca46a371e1b7

I ended up adding 1 day of padding to either end, to make sure we get a superset of the IDs we want.

Reply:
I believe this query is now unused, remove it?

Reply:
Oops, thanks, removed in a9611c644c8b382fddbea56d670c283840bfc810